Translated Test-Rezepte (#294)
reactjs/de.reactjs.org: content/docs/testing-recipes.md

let container = null;
beforeEach(() => {
  // Ein DOM-Element als Renderziel einrichten
  container = document.createElement("div");
  document.body.appendChild(container);
});

afterEach(() => {
  // Bereinigung beim Beenden
  unmountComponentAtNode(container);
  container.remove();
  container = null;
});

it("Rendert mit oder ohne Namen", () => {
  act(() => {

>     render(<Hallo />, container);

Reviewer: ```suggestion
    render(<Hello />, container);
```
We decided to translate only comments and plain text to keep it consistent. Not the code itself.

});
```

> Verschiedene DOM-Ereignisse und ihre Eigenschaften werden in [MDN](https://developer.mozilla.org/en-US/docs/Web/API/MouseEvent) beschrieben. Beachten Sie, dass Sie `{ bubbles: true }` in jedem von Ihnen erstellten Ereignis übergeben müssen, damit es den React-Listener erreicht, da React Ereignisse automatisch an den Stamm delegiert.

Reviewer: ```suggestion
Verschiedene DOM-Events und ihre Eigenschaften werden in [MDN](https://developer.mozilla.org/en-US/docs/Web/API/MouseEvent) beschrieben. Beachte, dass Du `{ bubbles: true }` in jedem von deinen erstellten Events übergeben musst, damit es den React-Listener erreicht, da React-Events automatisch an den Stamm (engl. root) delegiert werden.
```

> Hinweis:
>
> Die React Testing Library bietet einen [übersichtlicheren Helfer](https://testing-library.com/docs/dom-testing-library/api-events) für das Auslösen von Ereignissen.

---

### Timer {#timers}

> Ihr Code könnte zeitgeberbasierte Funktionen wie `setTimeout` verwenden, um mehr Arbeit in der Zukunft zu planen. In diesem Beispiel wartet ein Multiple-Choice-Feld auf eine Auswahl und rückt vor, wobei eine Zeitüberschreitung eintritt, wenn innerhalb von 5 Sekunden keine Auswahl getroffen wird:

Reviewer: ```suggestion
Dein Code könnte zeitgeberbasierte Funktionen wie `setTimeout` verwenden, um mehr Arbeit in der Zukunft zu planen. In diesem Beispiel wartet ein Multiple-Choice-Feld auf eine Auswahl und rückt vor, wobei eine Zeitüberschreitung eintritt, wenn innerhalb von 5 Sekunden keine Auswahl getroffen wird:
```

});
```

> Sie können gefälschte Timer nur in einigen Tests verwenden. Oben haben wir sie aktiviert, indem wir `jest.useFakeTimers()` aufgerufen haben. Der Hauptvorteil, den sie bieten, besteht darin, dass Ihr Test nicht wirklich fünf Sekunden auf die Ausführung warten muss, und Sie mussten den Komponentencode auch nicht nur zum Testen komplizierter machen.

Reviewer: ```suggestion
Du kannst Fake-Timer nur in einigen Tests verwenden. Oben haben wir sie aktiviert, indem wir `jest.useFakeTimers()` aufgerufen haben. Der Hauptvorteil, den sie bieten, besteht darin, dass dein Test nicht wirklich fünf Sekunden auf die Ausführung warten  und du den Komponentencode auch nicht nur zum Testen komplizierter machen musst.
```

---

### Snapshot-Testen {#snapshot-testing}

> Mit Frameworks wie Jest können Sie mit [`toMatchSnapshot` / `toMatchInlineSnapshot`](https://jestjs.io/docs/en/snapshot-testing) auch "Schnappschüsse" von Daten speichern. Mit diesen können wir die gerenderte Komponentenausgabe "speichern" und sicherstellen, dass eine Änderung daran explizit als Änderung am Snapshot festgeschrieben werden muss.

Reviewer: ```suggestion
Mit Frameworks wie Jest kannst Du mit [`toMatchSnapshot` / `toMatchInlineSnapshot`](https://jestjs.io/docs/en/snapshot-testing) auch "Schnappschüsse" von Daten speichern. Mit diesen können wir die gerenderte Komponentenausgabe "speichern" und sicherstellen, dass eine Änderung daran explizit als Änderung am Snapshot festgeschrieben werden muss.
```

});
```

Es ist in der Regel besser, spezifischere Behauptungen aufzustellen, als Snapshots zu verwenden. Diese Art von Tests beinhalten Implementierungsdetails, damit sie leicht brechen, und Teams können gegenüber Snapshot-Brüchen desensibilisiert werden. Das [selektive Simulieren einiger untergeordneter Komponenten](#mocking-modules) kann dazu beitragen, die Größe von Snapshots zu reduzieren und sie für die Codeüberprüfung lesbar zu halten.

---

### Mehrere Renderer {#multiple-renderers}

> In seltenen Fällen führen Sie möglicherweise einen Test für eine Komponente aus, die mehrere Renderer verwendet. Beispielsweise können Sie Snapshot-Tests auf einer Komponente mit `react-test-renderer` ausführen, die intern `render` von `react-dom` innerhalb einer untergeordneten Komponente verwendet, um einige Inhalte zu rendern. In diesem Szenario können Sie Aktualisierungen mit `act()`s umschließen, die ihren Renderern entsprechen.

Reviewer: ```suggestion
In seltenen Fällen führen Sie möglicherweise einen Test für eine Komponente aus, die mehrere Renderer verwendet. Beispielsweise kannst du Snapshot-Tests auf einer Komponente mit `react-test-renderer` ausführen, die intern `render` von `react-dom` innerhalb einer untergeordneten Komponente verwendet, um einige Inhalte zu rendern. In diesem Szenario können Sie Aktualisierungen mit `act()`s umschließen, die ihren Renderern entsprechen.
```

---

### Etwas fehlt? {#something-missing}

> Wenn ein gängiges Szenario nicht abgedeckt ist, teilen Sie uns dies bitte auf der [Problemverfolgungswebsite](https://github.com/reactjs/reactjs.org/issues) für die Dokumentation mit.

Reviewer: ```suggestion
Wenn ein gängiges Szenario nicht abgedeckt ist, teile uns dies bitte im [Issue-Tracker](https://github.com/reactjs/reactjs.org/issues) für die Dokumentation mit.
```
